src/main/java/org/opensearch/knn/validation/ParameterValidator.java

```diff
@@ -14,35 +14,89 @@
 import org.opensearch.common.Nullable;
 import org.opensearch.common.ValidationException;
 import org.opensearch.knn.index.Parameter;
+import org.opensearch.knn.index.util.EngineSpecificMethodContext;
+import org.opensearch.knn.index.util.KNNEngine;
 
 import java.util.ArrayList;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
+
+import static org.opensearch.knn.common.KNNConstants.METHOD_PARAMETER_EF_SEARCH;
 
 public final class ParameterValidator {
 
     /**
      * A function which validates request parameters.
      * @param validParameters A set of valid parameters that can be requestParameters can be validated against
      * @param requestParameters parameters from the request
-     * @return
      */
     @Nullable
     public static ValidationException validateParameters(
         final Map<String, Parameter<?>> validParameters,
         final Map<String, Object> requestParameters
     ) {
+        return validateParameters(validParameters, requestParameters, null, null);
+    }
 
+    /**
+     * A function which validates request parameters.
+     * @param validParameters A set of valid parameters that can be requestParameters can be validated against
+     * @param requestParameters parameters from the request
+     * @param knnEngine The KNN engine
+     * @param context The engine specific method context
+     */
+    @Nullable
+    public static ValidationException validateParameters(
+        final Map<String, Parameter<?>> validParameters,
+        final Map<String, Object> requestParameters,
+        final KNNEngine knnEngine,
+        final EngineSpecificMethodContext.Context context
+    ) {
+        validateNonNullParameters(validParameters);
+
+        if (requestParameters == null || requestParameters.isEmpty()) {
+            return null;
+        }
+
+        List<String> errorMessages = new ArrayList<>();
+        Set<String> checkedParameters = new HashSet<>();
+
+        checkEngineSpecificErrors(knnEngine, context, errorMessages, checkedParameters);
+        validateRequestParameters(validParameters, requestParameters, errorMessages, checkedParameters);
+
+        return buildValidationException(errorMessages);
+    }
+
+    private static void validateNonNullParameters(Map<String, Parameter<?>> validParameters) {
         if (validParameters == null) {
             throw new IllegalArgumentException("validParameters cannot be null");
         }
+    }
 
-        if (requestParameters == null || requestParameters.isEmpty()) {
-            return null;
+    private static void checkEngineSpecificErrors(
```

Comment:
The method is kept generic enough to have validations of any kind lets keep it as it is. There is a generic message on the `doToQuery` layer, that should be enough for users to identify what went wrong, right? am I missing something here? if not lets revert this

Reply:
Updated to no change in this file. And added query type in the error message. https://github.com/opensearch-project/k-NN/pull/1765/files#diff-6d0df17543181185170f65c116cdb3514d910e1557f012b7959611ae92678080R549